api/src/services/schema.ts

```diff
@@ -0,0 +1,290 @@
+		Joi.object({
+			collection: Joi.string(),
+			field: Joi.string(),
+			meta: Joi.any(),
+			schema: Joi.object({
+				default_value: Joi.any(),
+				max_length: [Joi.number(), Joi.string(), Joi.valid(null)],
+				is_nullable: Joi.bool(),
+			})
+				.unknown()
+				.allow(null),
+			type: Joi.string()
+				.valid(...TYPES, ...ALIAS_TYPES)
+				.allow(null),
+		})
+	),
+	relations: Joi.array().items(
+		Joi.object({
+			collection: Joi.string(),
+			field: Joi.string(),
+			meta: Joi.any(),
+			related_collection: Joi.any(),
+			schema: Joi.any(),
+		})
+	),
+});
+
+const deepDiffSchema = Joi.object({
+	kind: Joi.string().valid('N', 'D', 'E', 'A').required(),
+	path: Joi.array().items(Joi.string()),
+	lhs: Joi.any().when('kind', { is: 'N', then: Joi.optional(), otherwise: Joi.any() }),
+	rhs: Joi.any().when('kind', { is: 'D', then: Joi.optional(), otherwise: Joi.any() }),
+	index: Joi.any().when('kind', { is: 'A', then: Joi.number(), otherwise: Joi.optional() }),
+	item: Joi.any().when('kind', { is: 'A', then: Joi.any(), otherwise: Joi.optional() }),
+});
+
+const applyJoiSchema = Joi.object({
+	hash: Joi.string().required(),
+	diff: Joi.object({
+		collections: Joi.array().items(
+			Joi.object({
+				hash: Joi.string().when('diff.0.kind', { is: 'N', then: Joi.optional(), otherwise: Joi.any() }),
+				collection: Joi.string(),
+				diff: Joi.array().items(deepDiffSchema),
+			})
+		),
+		fields: Joi.array().items(
+			Joi.object({
+				hash: Joi.string().when('diff.0.kind', { is: 'N', then: Joi.optional(), otherwise: Joi.any() }),
+				collection: Joi.string(),
+				field: Joi.string(),
+				diff: Joi.array().items(deepDiffSchema),
+			})
+		),
+		relations: Joi.array().items(
+			Joi.object({
+				hash: Joi.string().when('diff.0.kind', { is: 'N', then: Joi.optional(), otherwise: Joi.any() }),
+				collection: Joi.string(),
+				field: Joi.string(),
+				related_collection: Joi.any(),
+				diff: Joi.array().items(deepDiffSchema),
+			})
+		),
+	}).required(),
+});
+
+export class SchemaService {
+	knex: Knex;
+	accountability: Accountability | null;
+
+	constructor(options: Omit<AbstractServiceOptions, 'schema'>) {
+		this.knex = options.knex || getDatabase();
+		this.accountability = options.accountability || null;
+	}
+
+	async snapshot(): Promise<Snapshot> {
+		if (this.accountability?.admin !== true) throw new ForbiddenException();
+
+		const currentSnapshot = await getSnapshot({ database: this.knex });
+
+		return currentSnapshot;
+	}
+
+	async apply(payload: { hash: string; diff: SnapshotDiff }): Promise<void> {
+		if (this.accountability?.admin !== true) throw new ForbiddenException();
+
+		const { error } = applyJoiSchema.validate(payload);
+		if (error) throw new InvalidPayloadException(error.message);
+
+		const currentSnapshot = await this.snapshot();
+
+		const snapshotWithHash = this.getHashedSnapshot(currentSnapshot);
+
+		if (payload.hash !== snapshotWithHash.hash) {
+			for (const diffCollection of payload.diff.collections) {
```

Comment:
Maybe we should move this bit to a new method `validateDiff` so it's easier to test in separation

Reply:
Done in [7217049](https://github.com/directus/directus/pull/14713/commits/72170492785f0d061aa11173bf1ccd45cd02fcb3). @rijkvanzanten Is this what you had in mind?

Reply:
@paescuj Thanks for making the changes! Thoughts on making `validateDiff` a void function like `validateSnapshot`? Just curious whether we should make them behave a tad more consistent (since both are validating stuff in schema service's methods)

Reply:
Thanks for your feedback, @azrikahar! Absolutely, we should 😃 I'm going to update it according to your suggestions 👍

Reply:
LGTM as is 👍🏻